### ISSUE TITLE
llvm error: cfi: ` Assertion 'isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#111216]
I tried this code:
`rustc treereduce.out -Clto -Zsanitizer=cfi -Copt-level=1 -Cdebuginfo=2 --crate-type lib`
```rust
pub fn main() {
    std::process::Command::new("");
}

```

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Meta

`rustc --version --verbose`:
```
eac35583d2ffb5ed9e564dee0822c9a244058ee0
```

<details><summary>Backtrace</summary>
<p>

```
rustc: /home/matthias/vcs/github/rust_debug_assertions/src/llvm-project/llvm/include/llvm/Support/Casting.h:579: decltype(auto) llvm::cast(From *) [To = llvm::MDNode, From = llvm::Metadata]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
[2]    3933595 IOT instruction  RUSTFLAGS="" ~/.rustup/toolchains/local-debug-assertions/bin/rustc  -Clto
```

</p>
</details>

**Comment by matthiaskrgr** [comment]
another one:
```rust
pub enum MaybeOwned<'a> {
    Borrowed(&'a isize),
}

pub struct Inv<'a> {
    x: &'a mut &'a isize,
}

pub trait IntoMaybeOwned<'a> {
    fn into_maybe_owned(self) -> MaybeOwned<'a>;
}

impl<'a> IntoMaybeOwned<'a> for Inv<'a> {
    fn into_maybe_owned(self) -> MaybeOwned<'a> {
        panic!()
    }
}

````

**Comment by rcvalle** [comment]
This seems to have been fixed in #113593.
